plugins/fuse/server/server_fuse.py

```diff
@@ -0,0 +1,492 @@
+        return fh
+
+    def release(self, path, fh):
+        """
+        Release an open file handle.
+
+        :param path: path within the fuse.
+        :param fh: an open file handle.
+        :returns: a file descriptor.
+        """
+        with self.openFilesLock:
+            if fh in self.openFiles:
+                with self.openFiles[fh]['lock']:
+                    if 'handle' in self.openFiles[fh]:
+                        self.openFiles[fh]['handle'].close()
+                        del self.openFiles[fh]['handle']
+                    del self.openFiles[fh]
+            else:
+                return super(ServerFuse, self).release(path, fh)
+        return 0
+
+    def destroy(self, path):
+        """
+        Handle shutdown of the FUSE.
+
+        :param path: always '/'.
+        """
+        events.trigger('server_fuse.destroy')
+        return super(ServerFuse, self).destroy(path)
+
+
+@atexit.register
+def unmountAll():
+    """
+    Unmount all mounted FUSE mounts.
+    """
+    for name in list(fuseMounts.keys()):
+        unmountServerFuse(name)
+
+
+def unmountServerFuse(name):
+    """
+    Unmount a mounted FUSE mount.  This may fail if there are open files on the
+    mount.
+
+    :param name: a key within the list of known mounts.
+    """
+    with fuseMountsLock:
+        entry = fuseMounts.pop(name, None)
+        if entry:
+            events.trigger('server_fuse.unmount', {'name': name})
+            path = entry['path']
+            # Girder uses shutilwhich on Python < 3
+            if shutil.which('fusermount'):
+                subprocess.call(['fusermount', '-u', os.path.realpath(path)])
+            else:
+                subprocess.call(['umount', os.path.realpath(path)])
+            if entry['thread']:
+                entry['thread'].join(10)
+            # clean up previous processes so there aren't any zombies
+            try:
+                os.waitpid(-1, os.WNOHANG)
+            except OSError:
+                # Don't throw an error; sometimes we get an
+                # errno 10: no child processes
+                pass
+
+
+def mountServerFuse(name, path, level=AccessType.ADMIN, user=None, force=False):
+    """
+    Mount a FUSE at a specific path with authorization for a given user.
+
+    :param name: a key for this mount mount.  Each mount point must have a
+        distinct key.
+    :param path: the location where this mount will be in the local filesystem.
+        This should be an empty directory.
+    :param level: access level used when checking which resources are available
+        within the FUSE.  This is ignored currently, but could be used if
+        non-readonly access is ever implemented.
+    :param user: the user used for authorizing resource access.
+    :param force: if True, all resources are available without checking the
+        user or level.
+    :returns: True if successful.  'present' if the mount is already present.
+        None on failure.
```

Comment:
This return API seems a little confusing. Are there conditions when you would want to call mountServerFuse() and know that it was already present rather than just knowing that the path is mounted?

Reply:
One of my ideas is that you could make a per-token or per-user mount and only expose files that are readable to that user.  In this case, you might have code that always mounts that in a predictable  spot per user, in which case knowing the mount is already present would be useful.

So far, I've only found actual utility it have the full-access mount.  Since this is accessible to the server code, and the server code already has full access to all files, making a user-limited mount would only be useful if the consuming process can't reach other local filesystem locations.

Reply:
would it make sense to have a public function that explicitly makes the presence check? I'm just concerned about code that uses this function needing to check for ```True``` and ```"present"``` and ```None```

Reply:
After taking a look.  I'm inclined to agree with @kotfic.  Making the presence check its own call and simplifying the return value of this function would seem to make for a cleaner API.

Reply:
Yes -- I added an `isServerFuseMounted` method in one of the later commits.